File: src/forge/session.py
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path

from forge.config import ForgeConfig, load_config
from forge.db import Database
from forge.router import LLMRouter
from forge.state_machine import Phase, transition

SESSIONS_DIR = Path.home() / ".forge" / "sessions"


@dataclass
class Session:
    id: str
    idea: str
    phase: Phase
    cycle: int
    max_cycles: int
    deploy_target: str | None
    workspace: Path
    db: Database
    router: LLMRouter
    config: ForgeConfig

    @classmethod
    def create(cls, idea: str, deploy_target: str | None = None) -> "Session":
        sid = str(uuid.uuid4())[:8]
        session_dir = SESSIONS_DIR / sid
        (session_dir / "workspace").mkdir(parents=True)
        (session_dir / "logs").mkdir()
        cfg = load_config()
        db = Database(session_dir / "session.db")
        # Use the sid we created, not the one db.create_session returns
        now = datetime.now(timezone.utc).isoformat()
        db.conn.execute(
            "INSERT INTO sessions (id, idea, phase, cycle, created_at, config_json) VALUES (?, ?, 'IDEATION', 0, ?, ?)",
            (sid, idea, now, "{}"),
        )
        db.conn.commit()
        if deploy_target:
            db.update_session(sid, deploy_target=deploy_target)
        return cls(
            id=sid, idea=idea, phase=Phase.IDEATION,
            cycle=0, max_cycles=cfg.max_cycles, deploy_target=deploy_target,
            workspace=session_dir / "workspace",
            db=db, router=LLMRouter(cfg.tier_models()), config=cfg,
        )

    @classmethod
    def load(cls, session_id: str) -> "Session":
        session_dir = SESSIONS_DIR / session_id
        if not session_dir.exists():
            raise FileNotFoundError(f"Session {session_id!r} not found")
        cfg = load_config()
        db = Database(session_dir / "session.db")
        row = db.get_session(session_id)
        if row is None:
            raise ValueError(f"Session {session_id!r} not in database")
        return cls(
            id=session_id, idea=row["idea"], phase=Phase(row["phase"]),
            cycle=row["cycle"], max_cycles=row["max_cycles"],
            deploy_target=row["deploy_target"],
            workspace=session_dir / "workspace",
            db=db, router=LLMRouter(cfg.tier_models()), config=cfg,
        )

    @classmethod
    def load_last(cls) -> "Session":
        if not SESSIONS_DIR.exists():
            raise FileNotFoundError("No sessions found")
        dirs = sorted(SESSIONS_DIR.iterdir(), key=lambda p: p.stat().st_mtime, reverse=True)
        if not dirs:
            raise FileNotFoundError("No sessions found")
        return cls.load(dirs[0].name)

    def advance_phase(self, next_phase: Phase) -> None:
        transition(self.phase, next_phase)
        self.phase = next_phase
        self.db.update_session(self.id, phase=next_phase.value)

    def increment_cycle(self) -> None:
        self.cycle += 1
        self.db.update_session(self.id, cycle=self.cycle)

File: src/session.ts
import * as fs from "fs";
import * as path from "path";
import * as os from "os";
import { randomUUID } from "crypto";
import { ForgeDb } from "./db.js";
import { LLMRouter } from "./router.js";
import { ForgeConfig, loadConfig, normalizeSkillConfig } from "./config.js";
import { Phase, transition } from "./stateMachine.js";
import { type MdCatalog, listToolCallModels, SUPPORTED_PROVIDERS } from "./modelsdev.js";
import { AutoSelector } from "./autoSelector.js";

export const SESSIONS_DIR = path.join(os.homedir(), ".forge", "sessions");

const PROVIDER_ENV_KEYS: Record<string, string> = {
  anthropic: "ANTHROPIC_API_KEY",
  openai:    "OPENAI_API_KEY",
  google:    "GOOGLE_API_KEY",
  groq:      "GROQ_API_KEY",
  mistral:   "MISTRAL_API_KEY",
};

/**
 * Configures the router's automatic model selector when auto-selection is enabled in the config.
 *
 * When `cfg.profile` is `"auto"`, `cfg.autoOverseer` is present, and `catalog` is provided, this sets
 * the router's auto-selector to an AutoSelector initialized with the overseer, configured priority,
 * the list of available model IDs discovered from `catalog`, and a logger that writes `AUTO_SELECT`
 * events to `db` for `sessionId`.
 *
 * @param router - The LLMRouter to configure
 * @param cfg - Forge configuration; `profile`, `autoOverseer`, and `priority` are used when enabling auto-selection
 * @param db - Database used to persist auto-selection log events
 * @param sessionId - Session identifier used as the owner of logged events
 * @param catalog - Optional model catalog used to discover available model IDs; required for enabling auto-selection
 */
function wireAutoSelector(
  router: LLMRouter,
  cfg: ForgeConfig,
  db: ForgeDb,
  sessionId: string,
  catalog?: MdCatalog,
): void {
  if (cfg.profile !== "auto" || !cfg.autoOverseer || !catalog) return;
  const activeProviders = SUPPORTED_PROVIDERS.filter(p => !!process.env[PROVIDER_ENV_KEYS[p]]);
  const providers = activeProviders.length ? activeProviders : SUPPORTED_PROVIDERS;
  const available = listToolCallModels(catalog, providers, true).map(({ model }) => model.id);
  const logFn = (msg: string) => db.logEvent(sessionId, "AUTO_SELECT", msg);
  router.setAutoSelector(new AutoSelector(cfg.autoOverseer, cfg.priority, available, logFn));
}

/**
 * Apply a persisted skills snapshot from a session database row to a base configuration.
 *
 * @param current - The base `ForgeConfig` to which persisted skills should be applied
 * @param row - Database row that may contain a `config_json` string holding a serialized config (expected to include a `skills` field)
 * @returns The `ForgeConfig` updated with normalized skills from the stored snapshot if parsing and normalization succeed, otherwise the original `current`
 */
function applySessionSkillSnapshot(current: ForgeConfig, row: Record<string, unknown>): ForgeConfig {
  const raw = typeof row["config_json"] === "string" ? row["config_json"] : "{}";
  try {
    const parsed = JSON.parse(raw) as Record<string, unknown>;
    const skills = normalizeSkillConfig(parsed["skills"]);
    return current.withSkills(skills);
  } catch {
    return current;
  }
}

export class Session {
  constructor(
    public id: string,
    public idea: string,
    public phase: Phase,
    public cycle: number,
    public maxCycles: number,
    public deployTarget: string | undefined,
    public workspace: string,
    public db: ForgeDb,
    public router: LLMRouter,
    public config: ForgeConfig,
  ) {}

  static create(
    idea: string,
    deployTarget?: string,
    sessionsDir = SESSIONS_DIR,
    workspace?: string,
    catalog?: MdCatalog,
    configOverride?: ForgeConfig,
  ): Session {
    const id = randomUUID().slice(0, 8);
    const sessionDir = path.join(sessionsDir, id);
    fs.mkdirSync(path.join(sessionDir, "logs"), { recursive: true });
    const resolvedWorkspace = workspace ?? path.join(sessionDir, "workspace");
    fs.mkdirSync(resolvedWorkspace, { recursive: true });
    const cfg = configOverride ?? loadConfig();
    const db = new ForgeDb(path.join(sessionDir, "session.db"));
    db.createSession(idea, id, JSON.stringify(cfg.toJson()));
    db.updateSession(id, { workspace: resolvedWorkspace });
    if (deployTarget) db.updateSession(id, { deploy_target: deployTarget });
    const router = new LLMRouter(cfg.tierModels(), catalog);
    wireAutoSelector(router, cfg, db, id, catalog);
    return new Session(
      id, idea, Phase.IDEATION, 0, cfg.maxCycles, deployTarget,
      resolvedWorkspace,
      db, router, cfg,
    );
  }

  static load(sessionId: string, sessionsDir = SESSIONS_DIR, catalog?: MdCatalog): Session {
    const sessionDir = path.join(sessionsDir, sessionId);
    if (!fs.existsSync(sessionDir)) throw new Error(`Session ${sessionId} not found`);
    const db = new ForgeDb(path.join(sessionDir, "session.db"));
    const row = db.getSession(sessionId);
    if (!row) throw new Error(`Session ${sessionId} not in database`);
    const cfg = applySessionSkillSnapshot(loadConfig(), row);
    const workspace = row["workspace"]
      ? String(row["workspace"])
      : path.join(sessionDir, "workspace");
    const router = new LLMRouter(cfg.tierModels(), catalog);
    wireAutoSelector(router, cfg, db, sessionId, catalog);
    return new Session(
      sessionId, String(row["idea"]), row["phase"] as Phase,
      Number(row["cycle"]), Number(row["max_cycles"]),
      row["deploy_target"] as string | undefined,
      workspace,
      db, router, cfg,
    );
  }

  static loadLast(sessionsDir = SESSIONS_DIR, catalog?: MdCatalog): Session {
    if (!fs.existsSync(sessionsDir)) throw new Error("No sessions found");
    const dirs = fs.readdirSync(sessionsDir)
      .filter(name => fs.statSync(path.join(sessionsDir, name)).isDirectory())
      .map(name => ({ name, mtime: fs.statSync(path.join(sessionsDir, name)).mtimeMs }))
      .sort((a, b) => b.mtime - a.mtime);
    if (!dirs.length) throw new Error("No sessions found");
    return Session.load(dirs[0].name, sessionsDir, catalog);
  }

  advancePhase(next: Phase): void {
    transition(this.phase, next);
    this.phase = next;
    this.db.updateSession(this.id, { phase: next });
  }

  incrementCycle(): void {
    this.cycle++;
    this.db.updateSession(this.id, { cycle: this.cycle });
  }
}
